man/yaml2json.1
.TH yaml2json "1" "December 15, 2023" "" "yaml2json manual"

.SH NAME
yaml2json - Converts YAML input to JSON/JSON5 output.

.SH SYNOPSIS
.B yaml2json
[-o|--output
.I file
] [-m|--mode
.I
mode
] [-w|--overwrite]
.br
[-p|--pretty] [-k|--check] [-a|--compact] [-t|--trailing-commas]
.br
[-s|--single-quotes] [--escape-slashes] [--escape-unicode]
.br
[-l|--line-break] [-V|--version] [-h|--help]
[
.I file ...
]
.RE

.SH DESCRIPTION
yaml2json converts YAML input to JSON/JSON5 output.

.SH ARGUMENTS
.B
.IP "file"
read the JSON/JSONC/JSON5 input from a file

.SH OPTIONS
.B
.IP "-o|--output file"
write the JSON output to a file
.B
.IP "-t|--trailing-commas"
insert trailing commas to arrays and objects
.B
.IP "-s|--single-quotes"
format single-quoted instead of double-quoted strings
.B
.IP "--escape-slashes"
escape slashes by by prefixing them with a backslash
.B
.IP "--escape-unicode"
escape multibyte Unicode characters with \u literals
.B
.IP "-l|--line-break"
append a line break to the JSON output
.B
.IP "-p|--pretty"
print the JSON output with line breaks and indented
.B
.IP "-V|--version"
print the version of the executable and exit
.B
.IP "-h|--help"
print the usage information and exit

.RE
If no input file is specified, it will be read from standard input.

.SH EXAMPLES
.B
.IP "yaml2json config.yaml -o config.json -lp"
Covert a file with prettifying and ensuring a trailing line break.
.B
.IP "cat config.yaml | yaml2json > config.json"
Covert a file minifying the result.

.SH BUGS
yaml2json's bug tracker is located at <https://github.com/prantlf/v-yaml2json/issues>

.SH DOCUMENTATION
See the website for documentation <https://github.com/prantlf/v-yaml2json#readme>

.SH AUTHOR
Currently maintained by Ferdinand Prantl <prantlf@gmail.com>


CHANGELOG.md
# Changes

## [0.3.1](https://github.com/prantlf/v-yaml2json/compare/v0.3.0...v0.3.1) (2023-12-15)

### Bug Fixes

* Include man pages in asset archives ([130b7e7](https://github.com/prantlf/v-yaml2json/commit/130b7e7df5cdde54a4346dbec27e1e15abdebeb5))

## [0.3.0](https://github.com/prantlf/v-yaml2json/compare/v0.2.0...v0.3.0) (2023-12-15)

### Features

* Add the man file ([ecff662](https://github.com/prantlf/v-yaml2json/commit/ecff662b365ec2f400aaf9c785e937e13f4366d6))

## [0.2.0](https://github.com/prantlf/v-yaml2json/compare/v0.1.0...v0.2.0) (2023-12-14)

### Features

* Add the linux arm64 support ([d824b2c](https://github.com/prantlf/v-yaml2json/commit/d824b2c6375d71de1f6b129a6d61569f73afa1f0))

## [0.1.0](https://github.com/prantlf/v-yaml2json/compare/v0.0.7...v0.1.0) (2023-12-12)

### Features

* Add arm64 build, cross-compile for Windows ([8353966](https://github.com/prantlf/v-yaml2json/commit/8353966a82e22605460bb9c54a64716a0ad9ab84))

## [0.0.7](https://github.com/prantlf/v-yaml2json/compare/v0.0.6...v0.0.7) (2023-10-24)

### Bug Fixes

* Build with updated dependencies ([799e1f2](https://github.com/prantlf/v-yaml2json/commit/799e1f266aecf55ca443d77bcd33fc59c9a0ff5d))

## [0.0.6](https://github.com/prantlf/v-yaml2json/compare/v0.0.5...v0.0.6) (2023-07-09)

### Features

* Allow to control escaping slashes and unicode characters ([3cc9032](https://github.com/prantlf/v-yaml2json/commit/3cc9032fe5ef3f063a639b3edff559663c15e561))

## [0.0.5](https://github.com/prantlf/v-yaml2json/compare/v0.0.4...v0.0.5) (2023-07-09)

### Features

* Allow single quoted strings in the output (starting JSON5) ([d16c17f](https://github.com/prantlf/v-yaml2json/commit/d16c17fae243ca3d146e54c2b9ed60d82f0c18b7))

## [0.0.4](https://github.com/prantlf/v-yaml2json/compare/v0.0.3...v0.0.4) (2023-06-25)

### Features

* Allow adding trailing commas to array and objects when stringifying ([bcce047](https://github.com/prantlf/v-yaml2json/commit/bcce047eaa7e854a2e62420d9951ed7c232d511c))

## [0.0.3](https://github.com/prantlf/v-yaml2json/compare/v0.0.2...v0.0.3) (2023-06-11)

### Features

* Parse the command-line arguments using cargs ([0f626a4](https://github.com/prantlf/v-yaml2json/commit/0f626a4b34593b9ca0ce59300c4c5a884ecef212))

## [0.0.2](https://github.com/prantlf/v-yaml2json/compare/v0.0.1...v0.0.2) (2023-06-10)

### Features

* Allow condensing command-line options and use the equal sign to separate values ([8e3df3c](https://github.com/prantlf/v-yaml2json/commit/8e3df3cd32508d983b858eb1c44992712744df95))
* Support negated options by uppercasing ([808119d](https://github.com/prantlf/v-yaml2json/commit/808119db7c0e9e3a9a276ba750241303375be02d))

## 0.0.1 (2023-06-09)

Initial release.


src/yaml2json.v
import os
import prantlf.cargs { Input, parse }
import prantlf.json { StringifyOpts, stringify }
import prantlf.yaml { parse_file, parse_text }

const version = '0.3.1'

const usage = 'Converts YAML input to JSON/JSON5 output.

Usage: yaml2json [options] [<yaml-file>]

  <yaml-file>           read the YAML input from a file

Options:
  -o|--output <file>    write the JSON output to a file
	-t|--trailing-commas  insert trailing commas to arrays and objects
	-s|--single-quotes    format single-quoted instead of double-quoted strings
	--escape-slashes      escape slashes by by prefixing them with a backslash
	--escape-unicode      escape multibyte Unicode characters with \\u literals
  -l|--line-break       append a line break to the JSON output
  -p|--pretty           print the JSON output with line breaks and indented
  -V|--version          print the version of the executable and exit
  -h|--help             print th usage information and exit

If no input file is specified, it will be read from standard input.

Examples:
  $ yaml2json config.yaml -o config.json -lp
  $ cat config.yaml | yaml2json > config.json'

struct Opts {
mut:
	output          string
	trailing_commas bool
	single_quotes   bool
	escape_slashes  bool
	escape_unicode  bool
	line_break      bool
	pretty          bool
}

fn convert() ! {
	opts, args := parse[Opts](usage, Input{ version: version })!

	src := if args.len > 0 {
		if args.len > 1 {
			return error('too many input files')
		}
		parse_file(args[0])!
	} else {
		input := os.get_raw_lines_joined()
		parse_text(input)!
	}

	mut dst := stringify(src, StringifyOpts{
		pretty: opts.pretty
		trailing_commas: opts.trailing_commas
		single_quotes: opts.single_quotes
		escape_slashes: opts.escape_slashes
		escape_unicode: opts.escape_unicode
	})
	if opts.line_break {
		dst += '\n'
	}
	if opts.output.len > 0 {
		os.write_file(opts.output, dst)!
	} else {
		print(dst)
	}
}

fn main() {
	convert() or {
		eprintln(err)
		exit(1)
	}
}
